lib/ui/text/paragraph_builder.cc

```diff
@@ -115,51 +134,126 @@ FOR_EACH_BINDING(DART_NATIVE_CALLBACK)
+
+  const uint8_t* uint8_data = static_cast<const uint8_t*>(byte_data.data());
+  uint8_t mask = uint8_data[0];
+
+  size_t byte_count = 1;
+  if (mask & sFontWeightMask) {
+    paragraph_style.strut_font_weight =
+        static_cast<txt::FontWeight>(uint8_data[byte_count++]);
+  }
+  if (mask & sFontStyleMask) {
+    paragraph_style.strut_font_style =
+        static_cast<txt::FontStyle>(uint8_data[byte_count++]);
+  }
+
+  float float_data[byte_data.length_in_bytes() - byte_count / 4];
+  memcpy(float_data, static_cast<const char*>(byte_data.data()) + byte_count,
+         byte_data.length_in_bytes() - byte_count);
+  size_t float_count = 0;
+  if (mask & sFontSizeMask) {
+    paragraph_style.strut_font_size = float_data[float_count++];
+  }
+  if (mask & sLineHeightMask) {
+    paragraph_style.strut_line_height = float_data[float_count++];
+  }
+  if (mask & sLeadingMask) {
+    paragraph_style.strut_leading = float_data[float_count++];
+  }
+  if (mask & sForceStrutHeightMask) {
+    // The boolean is stored as the last bit in the bitmask.
+    paragraph_style.force_strut_height = (mask & 1 << 7) != 0;
+  }
+
+  if (mask & sFontFamilyMask) {
+    return true;
+  }
+  return false;
+}
+
+ParagraphBuilder::ParagraphBuilder(
+    tonic::Int32List& encoded,
+    Dart_Handle strutData,
+    const std::string& fontFamily,
+    const std::vector<std::string>& strutFontFamilies,
+    double fontSize,
+    double lineHeight,
+    const std::u16string& ellipsis,
+    const std::string& locale) {
   int32_t mask = encoded[0];
   txt::ParagraphStyle style;
+
   if (mask & psTextAlignMask)
     style.text_align = txt::TextAlign(encoded[psTextAlignIndex]);
 
   if (mask & psTextDirectionMask)
     style.text_direction = txt::TextDirection(encoded[psTextDirectionIndex]);
 
-  if (mask & psFontWeightMask)
+  if (mask & psFontWeightMask) {
     style.font_weight =
         static_cast<txt::FontWeight>(encoded[psFontWeightIndex]);
+    style.strut_font_weight =
+        static_cast<txt::FontWeight>(encoded[psFontWeightIndex]);
+  }
 
-  if (mask & psFontStyleMask)
+  if (mask & psFontStyleMask) {
     style.font_style = static_cast<txt::FontStyle>(encoded[psFontStyleIndex]);
+    style.strut_font_style =
+        static_cast<txt::FontStyle>(encoded[psFontStyleIndex]);
+  }
 
-  if (mask & psFontFamilyMask)
+  if (mask & psFontFamilyMask) {
     style.font_family = fontFamily;
+    style.strut_font_families = strutFontFamilies;
+  }
 
-  if (mask & psFontSizeMask)
+  if (mask & psFontSizeMask) {
     style.font_size = fontSize;
+    style.strut_font_size = fontSize;
+  }
 
-  if (mask & psLineHeightMask)
+  if (mask & psLineHeightMask) {
     style.line_height = lineHeight;
+    style.strut_line_height = lineHeight;
+  }
 
+  if (mask & psStrutStyleMask) {
+    style.strut_enabled = true;
+    // Decode strut returns true if there is a font family string available.
+    if (decodeStrut(strutData, style)) {
+      style.strut_font_families = strutFontFamilies;
```

Comment:
Pass `strutFontFamilies` to `decodeStrut` so that `decodeStrut` can handle all strut fields including `style.strut_font_families`